### ISSUE TITLE
Collection of RPG_RT bugs for compatibility

### ISSUE THREAD

**Issue by fmatthew5876** [EasyRPG/Player#1486]
There are various RPG_RT bugs out there which may be depended on by different games. I've created this issue to catalog them. At some point we may want to enable a compatibility mode switch for some of these bugs if we find games need them.

Please list your known RPG_RT bugs here:

* Break statements in nested loops don't obey scope. They always jump to the next `EndLoop`
* Medicine usability is ignored in rm2k3 battles.

**Comment by CherryDT** [comment]
I have a huge list.

Maybe I should do what I wanted to do a while ago while discussing how to approach any possible future fixes: Open a GitHub repo for RM2k(3) but for issues only, similar to https://github.com/Microsoft/wsl

Thing is just that it's a huge amount of work.

**Comment by Ghabry** [comment]
iirc the "Target select" window (where you select a party member to use a item/skill on in the menu) ignores the "Half SP cost" attribute and shows the full SP cost but when executing the value is correct (halved).

**Comment by CherryDT** [comment]
> Flash saturation and Flash period (?) are not saved in the savegame.
> [Player/src/game_screen.h](https://github.com/EasyRPG/Player/blob/1d5aedccce4b2fbf0328bbb264b4631108cb9597/src/game_screen.h#L101)

Are you sure? It should be saved as SaveScreen.flash_current_level and SaveScreen.flash_time_left

**Comment by Ghabry** [comment]
> Are you sure

No Im not, just did a code search.

This bug here: #1391 
Negative E values which allows "enemy attacks can heal" is broken across all RPG Maker versions.

**Comment by fmatthew5876** [comment]
> 
> 
> > Flash saturation and Flash period (?) are not saved in the savegame.
> > [Player/src/game_screen.h](https://github.com/EasyRPG/Player/blob/1d5aedccce4b2fbf0328bbb264b4631108cb9597/src/game_screen.h#L101)
> 
> Are you sure? It should be saved as SaveScreen.flash_current_level and SaveScreen.flash_time_left

The easyrpg variables he refers to are for continuous flashing. You need copies of these 2 fields to tell the flasher how to repeat the flash. #1501

**Comment by fmatthew5876** [comment]
Rm2k3 doesn't let you put all party members in the back using the Row command. At least 1 member must be in the front.

However, if you put some party members in the back row and then remove the front row members, you'll end up with a party with everyone in the back.

**Comment by Ghabry** [comment]
![jrenlbf](https://user-images.githubusercontent.com/1331889/48969650-cd039e00-f001-11e8-9b55-ae9ef4319435.gif)

via https://rpgmaker.net/forums/topics/17752/?post=634525#post634525

**Comment by fmatthew5876** [comment]
All items that trigger skills use the source of the skill to the highest level member of the party (first one if same level).

That means that items that trigger skills that have usage == self will always be used on the highest level character in the party.

**Comment by fmatthew5876** [comment]
Not sure if this is really a bug:

If an actor has an "Attack All" weapon and gets confused or provoked, he will only attack one ally or enemy.

**Comment by fmatthew5876** [comment]
Create a weapon in 2k3 that can be used to activate a skill. Make it use the escape skill.

It will be usable from the menu. When you try to use it, it'll go to the actor select screen as if it's a healing item. Use it on a actor and you'll hear the escape skill's sound effect but nothing happens.

**Comment by fmatthew5876** [comment]
Create a skill with:
* Heals Death
* 0 power, phys, mag, var

Use it on a dead actor in battle in 2k3,

The actor will revive with 1hp, but the displayed green number will say -1 :exclamation:

**Comment by fmatthew5876** [comment]
Set a move route on the player which changes the player graphic. Then save and load the game.

Upon loading the hero graphic is reset.

**Comment by Tayruu** [comment]
[**Here's a bug**](https://forums.rpgmakerweb.com/index.php?threads/rm2k-scroll-map-offsets-from-player.106898/) I observed a while back around panning and player movement.

tldr: Panning the map away from the player, then moving the player into that new camera frame, causes the camera to become offset until the player returns to the tile they were originally on.

Though frankly I doubt there's any game that makes use of this behaviour to necessitate replicating it.

**Comment by fmatthew5876** [comment]
RPG Maker editor only lets you do a maximum of 10 terrain damage per step. Given that hp can go into the thousands in 2k3, this makes terrain damage all but useless as a game mechanic.

**Comment by fmatthew5876** [comment]
Item and Skill Usability bugs
--------------------------

* Normal Skills are usable in the menu only if they heal hp, mp, or at least one state which persists after battle. Items which invoke normal skills are usable if they heap hp, mp, or any state.
* 2k3: Equipment which invokes a skill is always usable in battle, regardless of the skills.
    * You can use this to invoke an escape or teleport in battle, which does nothing
    * You can use this to invoke a switch skill which would normally not be usable in battle
* 2k3: Equipment which invokes a skill is usable on the map if and only if it targets self or allies.
    *  This means for escape, teleport, and switch skills, they will only be usable in the menu if the scope is set to ally. This is not possible in editor unless you change them to normal type, change the scope, and then change them back to the original skill type.
    * This also means you can use switch skills normally only usable in battle.
* In the menu, Items which invoke skills require that at least one member of your party is alive and can use the item. This makes sense as the actor would have to "use" the item on another actor for them to receive the effect. However this check is only done for normal skills. Escape, teleport, and switch skills invoked by items are always usable even if nobody in the party could actually use the item.

**Comment by fmatthew5876** [comment]
Enemy AI Hidden Enemy Revive Bug
---------------------

Enemy AI will try to use revive skills on both dead and hidden enemies. This is because RPG_RT check for `Exists` instead of only checking for death.

This bug exists in 2k and 2k3.

**Comment by fmatthew5876** [comment]
2k Battle: Death + other states bug
----------------------------

Create a skill which inflicts death and poison. Set poison's rating to 100 so it's not removed with death. Now use this skill on an enemy.

The enemy with "die" twice. The bug is that the RPG_RT code repeats the "death" message and animation for all states inflicted by the skill.

**Comment by fmatthew5876** [comment]
2k3 Monster Blink Bug
-----------------

In 2k battle monsters do the blinking animation when they take damage, and then later fade out when they die.

In 2k3 because there are no messages or delays, when an enemy is killed it blinks and fades at the same time.

**Comment by fmatthew5876** [comment]
Revive effects bug
--------------------

Create a skill which raises hp, sp, atk, def, spi, agi and revives with 50% hit rate.

This skill will raise the parameters of dead characters, even if the revive part fails. 

This happens in 2k and 2k3, but is most obvious in 2k as you see the messages on the console.

**Comment by fmatthew5876** [comment]
Item Sp recovery on dead allies bug
----------------------

Medicine items which recover sp work on dead allies.

**Comment by CherryDT** [comment]
> ## Item Sp recovery on dead allies bug
> Medicine items which recover sp work on dead allies.

Not sure if that's a bug though... Death doesn't otherwise affect SP, so it would make sense for example to top up a dead allies SP with an item and then revive them with a skill (if my character with the reviving skill isn't ready first)

**Comment by Ghabry** [comment]
Is just not intuitive. I can't remember any other game I played where you can buff stats of dead actors.

So maybe it is a feature but it didn't age well 🤔

**Comment by fmatthew5876** [comment]
Skill Usability States Bug
--------------------

In Battle2k if you have a battler use a skill, and then something happens (example mp drain) which causes that skill to no longer be usable, the battler will do nothing that round.

This makes sense, but the way it's done skips checking for confuse, provoke, and processing states in general. The battler won't for example auto heal, or take poison damage, etc..

**Comment by fmatthew5876** [comment]
2k3 Battle Event triggering item action bug
-------------------

Whenever a battler does an action, the battle interpreter runs events (1) before action, (2) immediately after a switch is activated by and item or skill and (3) and after the action, allowing different page conditions in each case.


Due to a bug in the way they separated actor algo processing from enemies to implement CBA, step (3) does not occur when an actor uses an item.

Details in #2405

**Comment by fmatthew5876** [comment]
2k3 Battle Event Turns sequencing bug (Active ATB)
------------------------

Create battle event which happens on turn 1 and another that happens on turn 2 and has 1 enemy.

Right before the first battler acts, the turn will increment to 1, and the event will fire. Then the second battler, and the turn 2 event fires.

Now set atb to active, start the battle and enter the item or skill window. Stay in the window until the enemy atb fills up and they are ready to act, then perform the skill or item action.

What should happen:
Event pages reset, Turn 1 event fires, enemy acts, Event pages reset, Turn 2 event fires, actor acts

What does happen:
Event pages reset, Turn 2 event fires, enemy acts, actor acts

The reason is because the battle system incorrectly processes "next turn / reset pages" when a battler has a battle action scheduled and not when it first runs. So anytime you can trigger multiple actions to queue up like this, events will run only once before all the actions and reflect the last turn numbers from all the queued actions.

**Comment by fmatthew5876** [comment]
2k3 Battle Normal + First Strike vs Initiative / Surround differences
----------------------

In all of these battle conditions, the actors start with full ATB and enemies with empty ATB.

The first strike flag has an additional behavior in that the enemies will never do anything until at least one actor performs an action. For Initiative/Surround the enemies will attack you if you wait for their ATB to fill up.

The reason this works is because RPG_RT will still count up enemy ATB in first strike, but as soon as they act it will cancel their action. The first strike flag is cleared whenever an actor acts, so that allows enemies to act.

The RPG Maker editor forbids you from changing the first strike flag when the condition is not "Normal", however you can change the battle to normal, set the first strike flag, and then change it to initiative or surround to get the first strike behavior for those.

RPG_RT does clear the first strike flag for back and pincer attacks.

**Comment by fmatthew5876** [comment]
2k3 Battle enemy target branch command
---------------

The command has many bugs rendering it almost unusable. Details here: https://github.com/EasyRPG/Player/issues/2405#issuecomment-716298981

**Comment by fmatthew5876** [comment]
2k3 Battle Switch skill combo bug
----------------------

2k3 Switch skill edge case
-----------------------

Create a battle with 4 events, all activated by switch one, which print the page number and turn off the switch.

Crate one more event on turn 0 which enables a 3x combo for the hero for skill.

Finally, use a skill which activates switch 1.

What you would expect to see is "1", "2", "3". That is each time the skill is repeated the battle interpreter gets called again right after.

Instead due to some weird timing issues, you only get "1", "2"

**Comment by fmatthew5876** [comment]
2k3 negative attribute absorb numbers bug
----------------

Use a skill with a negative attribute that also aborbs on an enemy.

Instead of showing a green healing number, RPG_RT will show a negative number in white, looking like negative damage.

**Comment by fmatthew5876** [comment]
2k3 negative absorb
--------------

When an absorb skill effect is inverted by negative attributes, HP effect will heal the enemy but not affect the source. SP effect however will invert absorb, healing the enemy and hurting the source.

**Comment by Ghabry** [comment]
Invisible Gauge (Visual problem)
--------------

This affects battle 2k3 types with semi-transparent windows:

The gauge is drawn transparent but above the selection box. This means that the color of the gauge would be wrong because of the selection box as the draw order is "Background, Selection Box, Gauge, Text".

Instead of thinking about a solution they simply did not draw the gauge

**Comment by CherryDT** [comment]
I don't understand the issue, can you show a screenshot?

**Comment by Ghabry** [comment]
@CherryDT 
Aded a picture as you can see for transparent there is no gauge shown for the highlighted entry.

**Comment by CherryDT** [comment]
Thank you

**Comment by Ghabry** [comment]
Garbage data in BattleTest equipment
---------

Looks like the editor uses a 32 bit value for it but only initialized the lower 16 bit.

When I alter the battletest equipment the values saved are sometimes really high. The solution seems to be to strip of the upper 16 bit with "0xFFFF".

Needs a check what RPG_RT does. Maybe reads it directly in a 16 bit value which results in a truncation?

**Comment by Ghabry** [comment]
**No repositioning for CBA multi-attacks**
-----

When doing a CBA attack with a multiplier (x2 or x3) and a "Move Towards Enemy" the hero will only move to the first enemy. When this enemy dies the hero will hit the other enemies but does not move to them. Looks like lazy programming.

**Comment by Ghabry** [comment]
## CBA ranged attack with multiplier continues even when everything is dead
-------------

Preparation:

1. Set the CBA multiplier to x2 or x3
2. Set the CBA to Ranged
3. Enable in the Item effects (Item database tab) the "Attack all enemies"

Execution:

1. Start a battle test
2. Attack the enemies with your CBA weapon. It should be a one-hit kill.
3. After the 1st attack everything dies, but a 2nd attack is fully shown and a 3rd is interrupted by the victory animation :facepalm:

**Comment by Ghabry** [comment]
## Menu cursor does not flash in battle

When the command window containing Attack, Skill, Defend etc. is focussed the cursor does not animate.

Well, do not think supporting this is worth it ^^'

**Comment by Ghabry** [comment]
# Wrong cursor position when using Subset Skills

(Found by Dyluck):

Basically in RM2K3 after using a skill in battle, sometimes the cursor starts at a different skill on the next turn, instead of the last skill used. The reason is that the cursor will move down by a number determined by ALL skills the hero has, instead of what skills are in the current skill command window.  So for example if your database has these skills:

- Fire 1 (Magic)
- Fire 2 (Magic)
- Fire 3 (Magic)
- Heal 1 (Priest)
- Heal 2 (Priest)
- Heal 3 (Priest)

...and your hero has all 6 skills and both the Magic and Priest command. Then if you use Heal 2 in battle for example, the cursor will see this at the 5th entry of all your skills, and so it will move down 5 spots in every skill window. So in the Magic window, the cursor ends up pointing at Fire 3. In the Priest window, the cursor ends up on Heal 3.

**Comment by Ghabry** [comment]
## Escape/Teleport skill incorrect teleport behaviour

When these skills are used to change the map the behaviour appears to be similiar to the "Vehicle Teleport":

- Pictures are not removed
- Battle animations do not stop
- (to be tested: active events are not set to 0)

**Comment by Mimigris** [comment]
### Transparency overflow on a picture
Using a transparency value over 100 for pictures will result in the display of various colours serving as an overlay depending on the value instead of being transparent.

[See this video](https://www.youtube.com/watch?v=oxT0rsHAVjo&t=1m2s).

Used in some games, e.g. Helen's Mysterious Castle when beating a boss.

**Comment by Ghabry** [comment]
### Load Game event selects "Save Game" in menu

Reported by Dragoon Kamillo: I noticed strange thing in RM2k3 1.12, if you get load menu from event and closed this (this same if you really load any savefile) and after open menu - cursor will be set on save position. If you not have save option in normal menu this will be set OK in first position but one difference is visible moment without cursor in menu.

![2k3](https://github.com/user-attachments/assets/953a2b17-6a9b-4525-81b4-f19dd71653dd)

**Comment by Ghabry** [comment]
Behaviour of the Gold and the Message Window in Engine 2k3 < 1.10

- M = Message Window, G = Gold Window
- O = Opaque, T = Transparent, S = Semi-Transparent

The (!) is the one we don't implement correctly (see #3012)

| SysTransp | Battle | Message Opaque | Message Transparent |
|-----------|--------|----------------|---------------------|
| No        | A      | MO, GO         | MT, GO              |
| No        | B/C    | MO, GO         | MT, GO              |
| Yes       | A      | MO, GO         | MT, GO (!)          |
| Yes       | B/C    | MS, GS         | MT, GT              |

---

For Engine 2k3 >= 1.10 where the bug is fixed

| SysTransp | Battle | Message Opaque | Message Transparent |
|-----------|--------|----------------|---------------------|
| No        | A      | MO, GO         | MT, GO              |
| No        | B/C    | MO, GO         | MT, GO              |
| Yes       | A      | MS, GS         | MT, GT              |
| Yes       | B/C    | MS, GS         | MT, GT              |

---

2k: Does not support the message transparency setting

**Comment by Mimigris** [comment]
Documenting all of my finds that I got when I was updating the RPG Maker 2000 Test suite. A lot of these are mostly just "out of range values behaviour" that cannot necessarily be classifed as "RPG_RT bugs", but I think it should better be documented at least somewhere in any case. None of those are supported by the Player, also.

### Show Text, single use of `\`
While normally in a text, you can display a `\` character if you write it as `\\`, RPG_RT will be able to render a single `\` if it is the last character of a message box (can be from a written text or from the hero name).

### Change (Actor) Faceset, use of negative faces
If you specify to display the sprite 0 of a faceset, the sprite 4 will be displayed. With sprite -1, you can get sprite 3, with -2 you get 2. If you try to use a face with a value of -3 or lower, RPG_RT will throw an error.

### Number Input, out of range digits
If you specify a number of digits allowing you to type a value higher than the variable limit (e.g. 7~10 digits on 2000 8~10 digits on 2003), RPG_RT will accept it without capping the values.
If you specify 0 digits, no menu to input digits will open, meaning you cannot edit the variable, and pressing enter will make you input no value in the variable. If you specify a negative digit, it will be the same thing, but the message box arrow will not be displayed.

### Change Items/Change Party Member/Change Skills, 16-bit signed numbers
Those three commands use 16-bit signed numbers to process the the ID of the requested entity, and as such it will loop back if you put a value outside of that range (e.g. requesting ID 65537 will loop back to ID 1). Thing is, RPG_RT will still process it as being the ID requested until it gets added, meaning that it will not be able to check if you already have an item/member/skill of that ID, and will be able to add it even if it should not be able to. In the case of items and skills, as those menus are sorted by ID, and that RPG_RT technically considers it to be of the requested ID until addition, the new ID will be added at the start of the list if the ID requested is negative, or at the end of the list if positive, no matter if it is not the supposed placement of the ID. Attempting to remove using a value of the range will simply fail (RPG_RT will simply fail to find the requested ID).

### Change EXP/Change Level with a negative value will not adjust the level properly
Was already documented in the TestGame, but if you add negative experience or a negative level to a hero, and that the hero should level down due to it, RPG_RT will solely adjust the EXP without making the hero level down (will remove the EXP/put the EXP to the minimum EXP for the previous level, no level change). The opposite is also true, removing negative EXP or a negative level will not cause any level up. The only case when using those where a level change could happen is if even with the negative addition of EXP, for instance, you are still put at enough EXP to level up (can be done by using this glitch). Also, note that EXP change done by the Change Level command is always calculated based on the current level of the hero, not the calculated EXP (if you are level 24, add -1 level, then add 1 level, since you still are level 24 in this case, the second add level command will put you level 25, not level 24).

### Change Gold/Inn with a negative value does not cap money
Was already documented in the TestGame, but if you add a negative amount of money or remove a negative amount of money, RPG_RT will not cap money to the money range of [0;999999]. The same thing can be done if you set a Inn to require a negative amount of money.

### Damage Processing Variance >= 21 and lack of cap
If you use a Damage Processing command that has a Variance of 21 or more, and that deals positive damage before the Variance application, then it is possible to deal negative damages.
Damage Processing does not has any damage cap (only caps are the editor limitations), so you can deal up to 2147483647 damage, and the variable set as the result can also go up to 2147483647 instead of respecting the range. You can also overflow or underflow by having modifiers or the variance applied.

### Display of a hero name with half-spaces (`\_`)
If you display a message box that has a hero name in it, and that the hero name has an half-space in it, RPG_RT will not render the half-space, but will still take it into account for the display speed.

### Transitions
Invalid transitions (e.g. transition of ID 30) will all keep the display for a small time, then abruptly cut.
Transition values are processed on 8 bits, it loops at 256.
For the Change Transition command, negative values apply the same logic than for the positive range.
For the Transition command, instead, every negative value is considered to be the same than the transition of ID -1, the Default transition.

### Negative map IDs
While not possible without database editing, RPG_RT considers negative map IDs to be valid without any issue.

### Show Animation: This Event used on event IDs 10001~10004
If Show Animation: This Event is used on events whose IDs are between 10001 and 10004 (IDs conflicting with those of the player and of the vehicles), then the animation will be shown on the player/the vehicles, not on those event IDs.

### Shake Screen with a negative power OR speed
If a Shake Screen command is used with either a negative power or a negative speed, then the shake will start by the right instead of starting by the left.

### Speed increase when a music is constantly called with no wait in a parallel process
RPG_RT behaviour.

### Out of range parameters for music/sounds
RPG_RT seems to cap the balance and volume to the valid range, nothing to see here.
RPG_RT will freak out after a few seconds if you use a negative fade in or out, and has no problems with positive ones.
The minimum speed for all formats seems to be 1%.
The maximum speed doesn't seem to really be set for MIDI (can go really fast).
The maximum speed for wav files (both music and sound effects) is 625%, any value higher will be capped at 625%.
The maximum speed for mp3 files is 435% If you play an mp3 file at an higher speed, if said mp3 is not already playing, it will be played at 100%. If said mp3 file is already playing, its fade, volume and balance will be updated to those of the new command, while the speed will be left unaltered.

### Shared list for the Teleport and Escape skills
The Teleport and Escape points list is actually shared, Escape using the first entry of the list, Teleport using all others. Internally, each entry is assigned an ID, with Escape having the ID 0, and Teleport having the ID assigned to the map ID. The command Set Escape Location always adjusts the first entry in the list, while the command Set Teleport Location will adjust no matter the ID requested.
As such,
1) It is possible to use Set Teleport Location, and delete the entry of the map ID 0, which will remove the Escape point, and move the first teleport point to be the escape point (if there was no teleport point, the list will be empty and RPG_RT will throw an error upon opening the skill menu).
2) It is possible to use Set Teleport Location, and add a negative map ID, which will shift the escape point to the Teleport list, as the teleport point will be added to the list and ranged by ID. The Escape skill do not allow using any negative or null teleport point (even if negative maps can be valid, as mentioned earlier), so the entry will be greyed out.

### Out of range speeds for panoramas
Panoramas that have a speed between -31 and -9, or between 9 and 31 in a direction, will go at higher speeds (seemingly same logic than for in-range speeds).
Panoramas that have a speed between -62 and -32, or between 32 and 62 in a direction, will not move at all.
Panoramas that have a speed of 63 or higher in a direction will make RPG_RT throw an error (invalid floating point)

### Out of upper layer tiles from a Change Tile command
Those tiles seem to just use the sprite and data from the last upper layer sprite of the tileset.

### Battle commands in the overworld (2000, would have to test later for 2003)
It is possible to use battle commands in the overworld through file editing.
Control Variables: Enemy Data, Change Enemy HP, Change Enemy MP, Change Enemy State, Enemy Encounter will always trigger an error on RPG_RT due to an invalid enemy being referenced.
Abort Battle will make RPG_RT do a bugged transition which will then lead to its crash.
Show Animation (battle version, internally a different command) will make display any animation on screen nor play any sound, but if you set it to wait for finishing, it will still make you wait the actual time needed for the animation (also note that you cannot interrupt an overworld animation with a battle animation, same for the opposite).
Conditional Branch (battle version, internally a different command) will trigger an error on RPG_RT if you ask if an enemy can move (again, invalid enemy referenced). The other branches (switch status, variable status, actor can move) work exactly like they would in battle. This can even have a purpose, as the conditional branch "Actor can move" (aka "Actor not affected by a status preventing from acting in battle") has no direct equivalent in the non-battle Conditional Branch (closest would be having a conditional branch to check each status, which can be tedious depending on what you want to code).

Lastly, the Scroll Map command seems to have odd behaviours when using out of range speeds, but I cannot really see any logic in it so far.
